cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/RemoteValidationTask.java

```diff
@@ -0,0 +1,114 @@
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Task for executing validation requests remotely
+ */
+public class RemoteValidationTask implements RunnableTask {
+
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(Schema.class, new SchemaTypeAdapter())
+    .serializeNulls()
+    .create();
+  private static final Logger LOG = LoggerFactory.getLogger(RemoteValidationTask.class);
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    TaskSystemAppContext systemAppContext = context.getRunnableTaskSystemAppContext();
+    RemoteValidationRequest remoteValidationRequest = GSON.fromJson(context.getParam(), RemoteValidationRequest.class);
+    String namespace = remoteValidationRequest.getNamespace();
+    String serializedRequest = remoteValidationRequest.getSerializedRequest();
+    StageValidationRequest validationRequest;
+    try {
+      validationRequest = GSON.fromJson(serializedRequest,
+                                        StageValidationRequest.class);
+      validationRequest.validate();
+    } catch (JsonSyntaxException e) {
+      throw new IllegalArgumentException(String.format("Unable to decode request body %s", serializedRequest), e);
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Invalid stage config", e);
+    }
+
+
+    Map<String, String> arguments = Collections.emptyMap();
+
+    // Fetch preferences for this instance and namespace and use them as program arguments if the user selects
+    // this option.
+    if (validationRequest.getResolveMacrosFromPreferences()) {
+      try {
+        arguments = systemAppContext.getPreferencesForNamespace(namespace, true);
+      } catch (IllegalArgumentException iae) {
+        // If this method returns IllegalArgumentException, it means the namespace doesn't exist.
+        // If this is the case, we return a 404 error.
+        throw new IllegalArgumentException(String.format("Namespace '%s' does not exist", namespace), iae);
+      }
+    }
+
+    Map<String, String> finalArguments = arguments;
+    Function<Map<String, String>, Map<String, String>> macroFn =
+      macroProperties -> evaluateMacros(namespace, finalArguments, macroProperties, systemAppContext);
+    StageValidationResponse stageValidationResponse = ValidationUtils.validate(validationRequest,
+                                                                systemAppContext.createPluginConfigurer(namespace),
+                                                                macroFn);
+    context.writeResult(GSON.toJson(stageValidationResponse).getBytes());
+  }
+
+  private Map<String, String> evaluateMacros(String namespace, Map<String, String> programArgs,
+                                             Map<String, String> macroProperties,
+                                             TaskSystemAppContext systemAppContext) {
+    Map<String, MacroEvaluator> evaluators = ImmutableMap.of(
```

Comment:
can this be moved higher in the call stack so that this map is not being called per `evaluateMacros ` macro call? Basically looks like a static final map which will not change per call?

Reply:
Done. Changed local execution also to be consistent.